util/src/main/java/tc/oc/pgm/util/nms/v1_9/NMSHacks1_9.java

```diff
@@ -151,4 +194,593 @@ public void postToMainThread(Plugin plugin, boolean priority, Runnable task) {
     Object nmsServer = refl.getNMSServer(refl.getCraftServerHandle(bukkitServer));
     refl.addCallableToMainThread(nmsServer, Executors.callable(task));
   }
+
+  @Override
+  public void sendPacket(Player bukkitPlayer, Object packet) {
+    if (packet != null && bukkitPlayer != null) {
+      protocolManager.sendServerPacket(bukkitPlayer, (PacketContainer) packet);
+    }
+  }
+
+  private List<Player> getViewingPlayers(Entity entity) {
+    // TODO: use radius defined in spigot.yml / sportpaper.yml
+    List<Player> players = new ArrayList<>();
+    for (Entity nearbyEntity :
+        entity.getWorld().getNearbyEntities(entity.getLocation(), 64, 64, 64)) {
+      if (nearbyEntity instanceof Player && nearbyEntity.getEntityId() != entity.getEntityId()) {
+        players.add((Player) nearbyEntity);
+      }
+    }
+    return players;
+  }
```

Comment:
i have the feeling this can go very, very wrong.

Not getting the viewing players from a real place (ie: entity tracker) and instead *assuming* who can see the entity can and will crash older clients. Every time a player dies and they're put to bed, if you receive that packet in 1.7 while not seeing the player (eg: too far away, or the player was vanished or in obs), you'll crash. This is not reliable, a player may have just crossed into view but not have been sent to you yet, or may have moved away but not been removed yet.

Reply:
I should be able to use the reflection annotations for this. I'll rewrite this method to use that.